### ISSUE TITLE
Data are normalized 

### ISSUE THREAD

**Issue by jeanetteclark** [NCEAS/metadig-checks#457]
## Description

Optional check to see whether data are normalized.

This will be a tough one to implement, but I think we can come up with some creative approaches to determine whether data are normalized or not. Some things to look for:

- doesn't have obscene amount of columns
- doesn't have a ton of repeating values

## Priority

- FAIR: Optional

## Issues

TBD

## Procedure

TBD

**Comment by doulikecookiedough** [comment]
To determine if data are generally normalized, I suggest the following items to inspect (in addition to the original suggestions):
- Doesn't have duplicate columns
- Doesn't have duplicate rows

So in total, the `data-normalized` check would be successful for each data object if data:
- Doesn't have duplicate columns
- Doesn't have duplicate rows
- Doesn't have obscene amount of columns
- Doesn't have a ton of repeating values

This check only runs if the data object is a text delimited table, similar to how the `well-formed` and `variables-congruent` check is structured.

**Comment by mbjones** [comment]
@doulikecookiedough where are your test cases for validating the tool is working correctly? I might make some suggestions there.

**Comment by doulikecookiedough** [comment]
@mbjones I'm currently using `metadig-py` and altering existing test data locally, which is the quickest path so far because the data object and its sysmeta are already present. Here are my steps:

1. Drag and drop a check that I want to test into `testdata/checks` into my local `metadig-py` repo
2. Copy and paste a pytest in the `test_checks` module that checks a single check
    - ex. `test_run_check_datatable_variables_congruent`
3. Alter the text delimited table (.csv file) associated with the pytest to point to the check I want to test, and add a print statement:
    ```
   ...
   result_data = json.loads(result)
   print(result_data)
    ```
4. Run `pytest -s` and preview the results

So while we can test checks and suites now in `metadig-py`, it's not very "easy" to test **data checks** specifically because we require all the data's associated objects and sysmeta to be found in a `hashstore`. Even if we have the client and code ready, the user still has to set up a hashstore and store everything the check needs. Perhaps there should be a hashstore directly tied to `metadig-py` that is part of the repo?

So I think it would be extremely helpful to have a process like this in `metadig-py` to make it more friendly to users/testing:
- User manually downloads the dataset from the `adc` or `knb`
- User provides the path to the folder + calls a function
- We will parse the folder for the resource map (ex. `science-metadata.xml`)
- We upload the data objects to a the `metadig-py` hashstore, which is `.gitignore`d (so we never check in any data that gets uploaded there)
- We download the associated sysmeta from the data objects and upload that as well (because the `sysmeta` for data objects is not included when downloading datasets)
- `metadig-py` now has what it needs to execute data checks, we can run the entire suite or a single check.

**Comment by mbjones** [comment]
yeah, that testing approach is pretty chaotic. The package should use `pytest`. It should be possible to run the tests with a small test corpus that is configured automatically. You could hashstore for that if you don't want to actually store the test docs in a hashstore instance. Alternatively, it should be straightforward to initialize a small hashstore and copy a few files into it in an initialization fixture, which, once present, would not need to be recreated for that machine and could be used in all of the tests. 

Whether you should sink time into all of that is something I'll leave up to you and @jeanetteclark to discuss and decide, but these types of manual approaches are how we generate a lot of tech debt. Either way, we need another issue than this data normalization one for building out a sane test infrastructure.

**Comment by doulikecookiedough** [comment]
Currently, a temporary hashstore is created whenever pytest runs and initializes the small test corpus that's in `tests/testdata` - which is why it's much easier to adapt a pytest to do a different data check than it would be to set something up from scratch. I will sync up with @jeanetteclark to discuss how we may want to move forward.

**Comment by jeanetteclark** [comment]
also be on the look out in this check for:

- multiple tables in the same file
- very similar but not exactly the same column names (eg: temperature_1m, temperature_2m, temperature_3m)

**Comment by mbjones** [comment]
@doulikecookiedough I think the way to move forward is to look at your current list of 5 steps (numbered 1-5), and remove steps 1-4. You'll be left with step 5, to run `pytest -s`. If you arelady have a hashstore corpus, use it, and put any test content you need into it (or have your test fixtures create the content you need if its too big to check in to github). There should be no manual copy/paste/move/write a novel steps in there.
